src/uwtools/rocoto.py

```diff
@@ -183,12 +183,51 @@ def _add_task_dependency(self, e: Element, config: dict) -> None:
         :param config: Configuration data for this element.
         """
         e = SubElement(e, STR.dependency)
+        for key in config.keys():
+            tag, _ = self._tag_name(key)
+            if tag in [STR.taskdep, STR.datadep, STR.timedep]:
+                self._add_task_dependency_operand(e, config=config)
+            elif tag in [STR.streq, STR.strneq]:
+                self._add_task_dependency_strequality(e, config=config)
+            elif tag in [STR.True_, STR.False_]:
+                self._add_task_dependency_boolean(e, config=config)
+            else:
+                raise UWConfigError("Unhandled dependency type %s" % tag)
+
+    def _add_task_dependency_operand(self, e: Element, config: dict) -> None:
+        """
+        ???
+        """
         for key, block in config.items():
             tag, _ = self._tag_name(key)
             if tag == STR.taskdep:
                 self._set_attrs(SubElement(e, STR.taskdep), block)
+            elif tag == STR.datadep:
+                self._set_attrs(SubElement(e, STR.datadep), config={"attrs": block})
             else:
-                raise UWConfigError("Unhandled dependency type %s" % tag)
+                self._set_attrs(SubElement(e, STR.timedep), config={"attrs": block})
+
+    def _add_task_dependency_strequality(self, e: Element, config: dict) -> None:
+        """
+        ???
+        """
+        for key, block in config.items():
+            tag, _ = self._tag_name(key)
+            if tag == STR.streq:
+                self._set_attrs(SubElement(e, STR.streq), config={"attrs": block})
+            else:
+                self._set_attrs(SubElement(e, STR.strneq), config={"attrs": block})
+
+    def _add_task_dependency_boolean(self, e: Element, config: dict) -> None:
+        """
+        ???
+        """
+        for key, block in config.items():
+            tag, _ = self._tag_name(key)
+            if tag == STR.True_:
+                self._set_attrs(SubElement(e, STR.True_), config={"attrs": block})
+            else:
+                self._set_attrs(SubElement(e, STR.False_), config={"attrs": block})
```

Comment:
After taking a look at the Rocoto XML documentation the boolean operators ```True``` and ```False``` will need to be taken care of in the ENTITY section I believe.

Here is what the boolean operators look like when added to the workflow:
```
<?xml version="1.0"?>
<!DOCTYPE workflow
[
 ...
  <!ENTITY RUN_GSI "</true>">
]>

...

<dependency>
  &RUN_GSI;
</dependency>
```

There is no actual str literal ```True``` or ```False``` being added to the ```<dependency>``` element, only the ```value``` of boolean operator. 

Maybe adding a condition in the ```def _doctype()``` method would be the appropriate place for this:
```
if entities := self._config[STR.workflow].get(STR.entities):
    tags = (f'  <!ENTITY {key} "{val}">' for key, val in entities.items())
```

Reply:
I think setting `</true>` or `</false>` via an entity is just a shortcut for users who are manually creating their Rocoto workflow XML documents so that, rather than going down into the XML document to find and toggle the desired literal e.g. `</true>` or `</false>` tag, they can have a named entity (variable) that they can toggle at the top of the document. This would also support a DRY document: If they had the same sort of dependency in multiple tasks, they could control all of then by toggling a single entity.

But I think this is orthogonal to what you're currently doing, as users should -- as long as we continue trying to translate _all_ Rocoto functionality from UW YAML to Rocoto XML -- be able to specify a boolean dependency explicitly if they want to do that, and those would need to be rolled out by our code as actual `</true>` and `</false>` elements in the XML tree.

I think the excerpt from the Rocoto docs you show above would be supportable by making it possible to specify an entity reference as the `text` of a `<dependency>` element. That way, a user could specify e.g. `&RUN_GSI;` as the text, and then define an entity mapping `RUN_GSI` to `</true>` -- and we already have support for declaring arbitrary entities. I think "making it possible" means ensuring that the JSON Schema for UW YAML supports having something like a `text` setting for a `dependency` subtree in the Rocoto section of UW YAML. We might need to ensure that the user-supplied value regex-matches an entity reference, as I doubt the RELAX NG schema permits truly arbitrary text between `<dependency>` and `</dependency` -- it's probably expecting other `<>` elements, so whatever the user supplies as a value had better expand to that.

This is actually kind of tricky because hand-written XML could contain entity references _anywhere_, and if they expand to valid XML, they'd pass RELAX NG schema validation. But I doubt we have support in the Rocoto section of UW YAML for arbitrary entity references appearing pretty much anywhere. So maybe we don't want to support that. It'd be a good question for discussion.

Reply:
Maybe more explicit, I think it **is** valid to have e.g. `<dependency></true></dependency>` in Rocoto XML. The user could write that literally, or the `</true>` could be interpolated in via entity expansion like `<dependency>&RUN_GSI;</dependency>`.

It seems like Rocoto's goal here is to provide an easy way to turn pieces of functionality on and off, so maybe we need to figure out how the same goal could be achieved in UW YAML. For example, we could decide that every task or metatask block in the Rocoto section of UW YAML can have an optional `enabled` flag with a boolean value, which would default to `true` (`True` in Python) if not defined. Then, when a user generates a Rocoto XML document, tasks explicitly marked `enabled: false` would simply not appear in the XML output. A downside to that is that the user would have to use UW to regenerate the XML document if they wanted to enable that task; they could not (easily) just edit the XML, because the task wouldn't be there at all. There are more complicated things that could be done, too, involving automatically injecting control entities for all tasks, and updating all tasks with `<dependency>` elements that reference the associated entities and default to `</true>`. Those entities would have to be auto-generated by UW in a way that could never conflict with user-defined entities, and the `<dependency>` elements would have to be `<and>`ed with any other dependencies. It could be quite tricky.

Reply:
I'll have to see the final code, but two potential options for dealing with something like that are: 1. Add conditional logic on the receiver side to deal with inputs with different structure, or 2. Massage the input before passing it so that it conforms to what the receiver is expecting. The latter is the favored approach of the "confident programming" camp: Minimize conditional behavior by taking steps to ensure that invariants are maintained. The suggestion I made earlier, to construct a `{"attrs": ...}` object and passing it to a receiver that expects an `attrs` key follows this line of thought. Your call on what's best based on what you're actually dealing with (sometimes the conditional logic just _moves_ and doesn't disappear, and if it both moves and _multiplies_ it may not be a good solution), but it's something to consider. In general, it's always worth considering whether conditional logic could be replaced by introduction of and adherence to convention -- which is why I asked whether we should just require an `attrs` key in UW YAML any time an XML element has attributes, as that convention would make treatment of all elements consistent.